### DISCUSSION TITLE
Mouse and keyboard bluetooth not working on blueretro Dreamcast

### DISCUSSION THREAD

**Discussion by Predatorx21** [darthcloud/BlueRetro#243] [Q&A] [answered]
I managed to put together a blueretro with 2 outputs for Dreamcast. The problem is that it does not recognize my bluetooth mouse and keyboard (it is a 'Genius' brand).
Change the default setting of output 1 to "default mouse" and output 2 to "default gamepad / keyboard".
Peripheral icons appear in the Quake 3 game but the peripherals do not respond. That makes me think it's a compatibility issue.
I leave some images
![20220329_122806](https://user-images.githubusercontent.com/102832869/161346673-1682975e-154b-4192-a386-0e1e531cb147.jpg)
![20220326_042517](https://user-images.githubusercontent.com/102832869/161346732-31348ae9-b3ea-4c94-af79-85dac8652133.jpg)
![20220326_042459](https://user-images.githubusercontent.com/102832869/161346746-6f67ad86-1a13-48fb-8dba-610dee95a383.jpg)

**Comment by darthcloud** [comment]
You could flash the BlueRetro_pad_test.bin firmware and connect your ESP32 via USB on PC (do not connect to dreamcast) and get the log:
https://github.com/darthcloud/BlueRetro/wiki/Getting-BlueRetro-debug-logs-via-Serial-port-Windows-10

Copy log to txt file and attach here.

**Reply by Predatorx21** [reply]
Here I attach the txt file of the log.
[Blueretro Log.txt](https://github.com/darthcloud/BlueRetro/files/8405350/Blueretro.Log.txt)

**Reply by darthcloud** [reply]
You need to flash the pad test firmware, and then connect your device while the log is running.

**Reply by Predatorx21** [reply]
I already updated the firmware and got this log

[Blueretro log 2.txt](https://github.com/darthcloud/BlueRetro/files/8405664/Blueretro.log.2.txt)

**Reply by darthcloud** [reply]
Your log show its running the DC firmware and their is no trace of a device trying to connect.

**Reply by Predatorx21** [reply]
So, what model of Bluetooth keyboard and mouse do you recommend that works with Blueretro?

**Reply by darthcloud** [reply]
Logitech

**Reply by Predatorx21** [reply]
I have one last question...Do Bluetooth 5.0 keyboards work on Blueretro?

**Reply by darthcloud** [reply]
BT 5.0 don't really means anything (beside faster BLE or maybe LE Audio) but I guess it's some marketing BS to say Bluetooth Low Energy (BLE) which was introduce in BT 4.0

BlueRetro support both Classic and LE Bluetooth and should support any HID devices that exist in theory.

In practice, interoperability issue may exist between the ESP32 and some device at the RF level.
Bugs may exist in the HID device support of BlueRetro that may get triggered by "weird" device.

Cheap device often got those weird implementation.

I haven't had any issue with brand name like Logitech.

**Reply by Predatorx21** [reply]
Hello Darthcloud. I tell you that I bought a Logitech BT mouse and the truth is that it works very well ! 

I also bought a 'Redragon' BT keyboard (logitech keyboards are very expensive), but no luck with that. The led flashes but never finishes pairing.
Too bad there is no firmware update to fix that problem with those keyboards. 

This is the problematic keyboard 👇

![20220408_163351.jpg](https://user-images.githubusercontent.com/102832869/162526779-b08ef86b-a6bd-4ba2-ac07-80b1b9fd33b9.jpg)![20220408_163427.jpg](https://user-images.githubusercontent.com/102832869/162526812-0cfac33c-9413-486e-ba0e-2fc3276a6530.jpg)

**Reply by darthcloud** [reply]
Its probably not much that is missing for them to work.

But I would need minimally the log of it trying to connect on BlueRetro. Or if its more complex issues I would need the keyboard.

**Reply by Predatorx21** [reply]
Well ! That gives me hope !

Here I got this log. I leave it attached.

[Blueretro Log_ Draconic K530RGB.txt](https://github.com/darthcloud/BlueRetro/files/8455945/Blueretro.Log_.Draconic.K530RGB.txt)

**Reply by darthcloud** [reply]
Their is an error in parsing one of the HID report but it's an useless one so it should not be the issue.

I think the KB doesn't like one of the BLE read response BlueRetro is giving.

It's not something easy to troubleshoot remotely unfortunately.

**Reply by Predatorx21** [reply]
uhh, what a bad news.

Does that mean you need the keyboard to fix the problem? 

Well, if you discover a possible solution please let me know, otherwise i will have to return the keyboard

**Reply by darthcloud** [reply]
Yes.

**Reply by Predatorx21** [reply]
Hello Darthcloud.
In the end it seems to me that I am going to return the BT Draconic keyboard 😕

What do you think about converting a USB keyboard to BT with Arduino Uno and the BT module?. Will it really work on Blueretro?

![Screenshot_20220420-123516_cropped.png](https://user-images.githubusercontent.com/102832869/164270450-ef5eb785-5b36-4a21-a3f5-6e5bcf7b3af8.png)

![Screenshot_20220420-123414_cropped.png](https://user-images.githubusercontent.com/102832869/164270494-c9d0c84f-99e4-4786-8923-3998fc6d3ff1.png)

**Comment by Predatorx21** [comment]
This is the model of the keyboard that I can not link...

![20220403_164016.jpg](https://user-images.githubusercontent.com/102832869/161445350-3839f218-8b2f-40e3-b238-0f085850b49b.jpg)

**Comment by darthcloud** [comment, marked as answer]
As a said before BlueRetro support both Classic and LE Bluetooth and should support any HID devices that exist in theory.

Adding support to any device is simply a matter of me having the device on hand and fixing the bugs in BlueRetro BT stack.

I made an amazon Wishlist of the most popular BT devices, if you feel generous buy me one and I'll make sure to make it work.
https://www.amazon.ca/hz/wishlist/ls/2RMFK6DHX29NQ?ref_=wl_share

Also let me know if I should add a specific device on the list.
